src/coreclr/nativeaot/Runtime/amd64/MiscStubs.asm

```diff
@@ -9,6 +9,18 @@ EXTERN memcpyGCRefsWithWriteBarrier : PROC
 EXTERN memcpyAnyWithWriteBarrier    : PROC
 EXTERN RhpGetThreadStaticBaseForTypeSlow : PROC
 
+EXTERN __guard_check_icall_fptr : QWORD
+EXTERN __guard_dispatch_icall_fptr : QWORD
+
+LEAF_ENTRY __guard_check_icall_fptrHack, _TEXT
+    jmp qword ptr [__guard_check_icall_fptr]
+LEAF_END __guard_check_icall_fptrHack, _TEXT
+
+LEAF_ENTRY __guard_dispatch_icall_fptrHack, _TEXT
+    jmp qword ptr [__guard_dispatch_icall_fptr]
```

Comment:
`__guard_dispatch_icall_fptr` is actually a pointer to the address of the helper, not the address of the helper (which kind of makes sense now that I think about it, since the helper is provided by Windows). We probably don't want to go through this jump thunk. This is my fault for not realizing it earlier. I this being an address of an address won't complicate things too much.

Reply:
I believe this should already just work on the JIT side as the JIT-EE interface can return helper addresses either as direct addresses or as one that needs to be accessed through an indirection.

I assume this cell ends up as read-only when mapped? Just verifying that we do not introduce an intermediate layer somewhere and end up defeating the purpose of the feature.

Reply:
> I believe this should already just work on the JIT side as the JIT-EE interface can return helper addresses either as direct addresses or as one that needs to be accessed through an indirection.

Ah. Works like charm! Pushed out another commit that does that. Thanks!

> I assume this cell ends up as read-only when mapped? Just verifying that we do not introduce an intermediate writeable layer somewhere and end up defeating the purpose of the feature.

Yes, IIRC the loader even validates this.